desc/integrals/_bounce_utils.py

```diff
@@ -700,127 +845,315 @@ def interp_to_argmin_hard(h, points, knots, g, dg_dz, method="cubic"):
+        expense of noisier gradients - noisier in the physics sense (unrelated
+        to the automatic differentiation).
+    upper_sentinel : float
+        Something larger than g. Choose value such that
+        exp(max(g)) << exp(``upper_sentinel``). Don't make too large or numerical
+        resolution is lost.
+
+    Warnings
+    --------
+    Recall that if g is small then the effect of β is reduced.
+    If the intention is to use this function as argmax, be sure to supply
+    a lower sentinel for ``upper_sentinel``.
 
     Returns
     -------
-    fig, ax
-        Matplotlib (fig, ax) tuple.
+    h : jnp.ndarray
+        Shape (..., num well).
 
     """
-    fig, ax = plt.subplots()
-    legend = {}
-    if include_knots:
-        for knot in ppoly.x:
-            _add2legend(
-                legend,
-                ax.axvline(
-                    x=knot, color="tab:blue", alpha=knot_transparency, label="knot"
-                ),
-            )
+    ext, g_ext = _get_extrema(knots, g, dg_dz, sentinel=0)
+    # Our softargmax(x) does the proper shift to compute softargmax(x - max(x)),
+    # but it's still not a good idea to compute over a large length scale, so we
+    # warn in docstring to choose upper sentinel properly.
+    argmin = softargmax(
+        beta * _where_for_fft_argmin(points, ext, g_ext, upper_sentinel),
+        axis=-1,
+    )
+    return jnp.linalg.vecdot(
+        argmin,  # shape is (..., num well, num extrema)
+        interp_rfft2(
+            T.eval1d(ext),
+            ext,
+            h[..., jnp.newaxis, :, :],
+            domain1=(0, 2 * jnp.pi / NFP),
+            axes=(-1, -2),
+        )[..., jnp.newaxis, :],
+        # adding axis to broadcast with num well axis
+    )
 
-    z = jnp.linspace(
-        start=setdefault(start, ppoly.x[0]),
-        stop=setdefault(stop, ppoly.x[-1]),
-        num=num,
+
+# TODO: Generalize this beyond ζ = ϕ or just map to Clebsch with ϕ.
+def get_fieldline(alpha_0, iota, num_transit, period):
+    """Get sequence of poloidal coordinates A = (α₀, α₁, …, αₘ₋₁) of field line.
+
+    Parameters
+    ----------
+    alpha_0 : float
+        Starting field line poloidal label.
+    iota : jnp.ndarray
+        Shape (num rho, ).
+        Rotational transform normalized by 2π.
+    num_transit : float
+        Number of ``period``s to follow field line.
+    period : float
+        Toroidal period after which to update label.
+
+    Returns
+    -------
+    fieldline : jnp.ndarray
+        Shape (num rho, num transit) or broadcastable.
+        Sequence of poloidal coordinates A = (α₀, α₁, …, αₘ₋₁) that specify field line.
+
+    """
+    # Δϕ (∂α/∂ϕ) = Δϕ ι̅ = Δϕ ι/2π = Δϕ data["iota"]
+    return alpha_0 + period * (
+        iota if iota.size == 1 else iota[:, jnp.newaxis]
+    ) * jnp.arange(num_transit)
+
+
+def fourier_chebyshev(theta, iota, alpha, num_transit):
+    """Parameterize θ along field lines ``alpha``.
+
```

Comment:
a little more description of the parametrization would be good? that it is some set of chebyshev polynomials enforced to only be piecewise continuous?

Reply:
The returned piecwise chebyshev series is $C^{\infty}$ everywhere. I only enforce it to be continuous to improve convergence as discussed in the notes, and in case someone needs the single valued version along field lines for secular integrals.

Thoughts on the description in ``Returns``?

Reply:
ah ok I see I think. It is a set of 1D chebyshev along different alphas, but if one were to add 2pi to them they would become continous across the alphas since it is the same field line just followed around for zeta > 2pi, with cuts made at the zeta=0 plane separating the different polynomials. Thanks